### ISSUE TITLE
Request for Feature: Reconfigure

### ISSUE THREAD

**Issue by nickman** [tinylog-org/tinylog#100]
The docs are fairly plain on this:

> The configuration of tinylog is immutable, and as soon as the first log entry is issued, further configuration changes will be ignored.

I would like to request an override on this. I use tinylog in a JavaAgent and really need to be able to rollup a new configuration and enable it at runtime because I just don't have enough information early on to commit to a configuration. Or put another way, it would be complicated to use some other form of logging until I do, and then commit to tinylog.

If this seems onerous, can you point me in the right direction for a reflective workaround where I can reset the configuration and reinitialize it ? I am confident I can do it at a point where I can assure no other logging activity will be happening so I am not opposed to this.

Thanks for this library. It's been a lifesaver for the JavaAgent I'm working on !

**Comment by pmwmedia** [comment]
Reconfiguration at runtime is a very rarely used feature of tinylog 1.x. Therefor and for a great [performance boost](http://tinylog.org/v2/benchmark/), I have droppt this feature in tinylog 2.x.

@nickman However, there might be a solution for your use case. Firstly, I have really to understand what you are doing in your JavaAgent. What is the source of your logging configuration? Which properties do you reconfigure at runtime?

**Comment by nickman** [comment]
Thanks for your response !
I decided to implement tinylog because my target systems where I want to
use the JavaAgent have a diversity of logging packages and I have found
that shading logging packages is a fiasco, so tinylog fits the bill
perfectly and if I come across a situation where a target system uses
tinylog, it's a very shading friendly package so overall it's working well
for me.
The JavaAgent has a complicated startup and I need to use the logger at
startup time so the logger starts with a "bootstrap" configuration.
However, at some point, the agent loads an XML configuration file which
contains segment like this:

*<logging>*
*   writer = file *
*   writer.file = log.txt *
*   writer.level = info*
*   <... more config ...>*
*</logging>*

At this stage, I want to read in the content of that XML block, convert to
Properties and re-configure the loggers to the "runtime" configuration.
I can ensure that there is only one thread running at the time, so I am not
too worried about concurrent access.
That's about it.

Thanks again !

//Nicholas

**Comment by pmwmedia** [comment]
You can create a custom reconfigurable [LoggingProvider](https://github.com/pmwmedia/tinylog/blob/v2.0/tinylog-api/src/main/java/org/tinylog/provider/LoggingProvider.java):

```java
package org.sample;

import java.lang.reflect.Method;
import java.util.Map;

import org.tinylog.Level;
import org.tinylog.configuration.Configuration;
import org.tinylog.core.TinylogLoggingProvider;
import org.tinylog.provider.ContextProvider;
import org.tinylog.provider.LoggingProvider;

public class ReconfigurableLoggingProvider implements LoggingProvider {
	
	private TinylogLoggingProvider realProvider = new TinylogLoggingProvider();
	
	public void reload() throws InterruptedException, ReflectiveOperationException {
		realProvider.shutdown();
		
		Method method = Configuration.class.getDeclaredMethod("load");
		method.setAccessible(true);
		Configuration.replace((Map) method.invoke(null));
		
		realProvider = new TinylogLoggingProvider();
	}

	@Override
	public ContextProvider getContextProvider() {
		return realProvider.getContextProvider();
	}

	@Override
	public Level getMinimumLevel() {
		return Level.TRACE;
	}

	@Override
	public Level getMinimumLevel(final String tag) {
		return Level.TRACE;
	}

	@Override
	public boolean isEnabled(final int depth, final String tag, final Level level) {
		return realProvider.isEnabled(depth + 1, tag, level);
	}

	@Override
	public void log(final int depth, final String tag, final Level level, final Throwable exception, final Object obj, final Object... arguments) {
		realProvider.log(depth + 1, tag, level, exception, obj, arguments);
	}

	@Override
	public void log(final String loggerClassName, final String tag, final Level level, final Throwable exception, final Object obj, final Object... arguments) {
		realProvider.log(loggerClassName, tag, level, exception, obj, arguments);
	}

	@Override
	public void shutdown() throws InterruptedException {
		realProvider.shutdown();
	}

}
```

Logging providers have to be registered as a service.

src/main/resources/META-INF/services/org.tinylog.provider.LoggingProvider:

```
org.sample.ReconfigurableLoggingProvider
```

Now you can define your logging provider in your tinylog.properties:

```properties
provider      = reconfigurable
writer        = console
writer.format = {date} {class}.{method}: {message}
```

I have also attached my tested sample project. There you can edit `tinylog.properties` at runtime, and you will see the new configuration immediately: [sample.zip](https://github.com/pmwmedia/tinylog/files/3154581/sample.zip)

**Comment by nickman** [comment]
Thanks ! Starting on this today.

**Comment by zulot** [comment]
Hi,
I tried to add, removed tagg on run whith last version off tinylog.
I used your example, but they doesn't work.

i see a value frozen who lock change.
i add code for change this value, but every day, if a new log come with tag, in the same day, i can't log on file this tag.

after i see this documentation : https://tinylog.org/v2/extending/#custom-logging-provider
But it is very different.

What is the best way for change during run the possibility to log or not log specific tag ?
